File: src/com/offz/minecraft/staminaclimbing/plugin/Stamina/StaminaBar.java
package com.offz.minecraft.staminaclimbing.plugin.Stamina;

import com.offz.minecraft.staminaclimbing.plugin.StaminaAndClimbing;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.boss.BarColor;
import org.bukkit.boss.BarStyle;
import org.bukkit.boss.BossBar;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerMoveEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.server.PluginDisableEvent;

import java.util.*;

public class StaminaBar implements Listener {

    public static List<UUID> toggled = new ArrayList<>();

    public static Map<UUID, BossBar> registeredBars = new HashMap<>();
    public void removeBossBarProgress(double amount, Player p, BossBar b){
        double progress = b.getProgress();
        if(progress - amount >= 0) {
            b.setProgress(b.getProgress() - amount);
        }else {
            b.setProgress(0);
        }
    }
    @EventHandler
    public void onDisable(PluginDisableEvent e){
        for (BossBar b : registeredBars.values()) {
            b.removeAll();
        }
    }
    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent e){
        if(!StaminaBar.toggled.contains(e.getPlayer().getUniqueId())) return;
        registerBar(e.getPlayer());
    }

    @EventHandler
    public void onPlayerQuit(PlayerQuitEvent e){
        if(!StaminaBar.toggled.contains(e.getPlayer().getUniqueId())) return;
        registeredBars.get(e.getPlayer().getUniqueId()).removeAll();
        registeredBars.remove(e.getPlayer().getUniqueId());
    }

    @EventHandler
    public void onPlayerMove(PlayerMoveEvent e) {
        if(!StaminaBar.toggled.contains(e.getPlayer().getUniqueId())) return;
        Player p = e.getPlayer();
        if(registeredBars.get(p.getUniqueId()) == null){
            registerBar(e.getPlayer());
        }
        BossBar b = registeredBars.get(p.getUniqueId());
        if (p.isSprinting() && b.getProgress() >= 0.1 + 0.0035) {
            removeBossBarProgress(0.0035, p, b);
        }else if (p.isSneaking()) {
        } else {
            removeBossBarProgress(0.00025, p, b);
        }
    }
    @EventHandler
    public void onPlayerFall(EntityDamageEvent e) {
        if (e.getEntity() instanceof Player && e.getCause() == EntityDamageEvent.DamageCause.FALL) {
            Player p = (Player) e.getEntity();
            BossBar b = registeredBars.get(p.getUniqueId());

            removeBossBarProgress((p.getFallDistance() * (p.getFallDistance()/2)) / 112.5, p, b);
        }
    }
    @EventHandler
    public void onPlayerDeath(PlayerDeathEvent e){
        if(!StaminaBar.toggled.contains(e.getEntity().getUniqueId())) return;
        Player p = e.getEntity();
        BossBar b = registeredBars.get(p.getUniqueId());
        b.setProgress(1);
    }
    public static void registerBar(Player p){
        BossBar b = Bukkit.createBossBar(ChatColor.BOLD+"Stamina", BarColor.BLUE, BarStyle.SEGMENTED_10);
        b.addPlayer(p);
        registeredBars.put(p.getUniqueId(), b);
    }
}

File: src/main/kotlin/com/mineinabyss/staminaclimb/stamina/StaminaBar.kt
package com.mineinabyss.staminaclimb.stamina

import com.github.shynixn.mccoroutine.bukkit.launch
import com.mineinabyss.idofront.entities.toPlayer
import com.mineinabyss.idofront.textcomponents.miniMsg
import com.mineinabyss.staminaclimb.*
import com.mineinabyss.staminaclimb.climbing.ClimbBehaviour
import com.mineinabyss.staminaclimb.modules.stamina
import com.mineinabyss.staminaclimb.nms.Tags
import kotlinx.coroutines.delay
import net.kyori.adventure.bossbar.BossBar
import org.bukkit.Bukkit
import org.bukkit.GameMode.ADVENTURE
import org.bukkit.GameMode.SURVIVAL
import org.bukkit.Location
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageEvent
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.PlayerGameModeChangeEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerMoveEvent
import org.bukkit.event.player.PlayerQuitEvent
import java.util.*
import kotlin.math.pow
import kotlin.time.Duration.Companion.seconds

object StaminaBar : Listener {
    internal var conf = stamina.config
    private val disabledPlayers: MutableList<UUID> = mutableListOf() //TODO persist

    @PublishedApi
    internal val registeredBars: MutableMap<UUID, BossBar> = mutableMapOf()
    private val velocities: MutableMap<UUID, Double> = mutableMapOf()
    private val barProgressTracker = mutableMapOf<UUID, Float>()

    fun climbEnabled(player: Player) = player.uniqueId !in disabledPlayers

    fun setClimbEnabled(player: Player, enable: Boolean) {
        if (climbEnabled(player) == enable) return
        val uuid = player.uniqueId
        if (enable) {
            disabledPlayers.remove(uuid)
            registerBar(player).progress(0f)
        } else {
            disabledPlayers.add(uuid)
            unregisterBar(uuid)
            player.stopClimbing()
        }
    }

    fun registerBar(player: Player): BossBar {
        val uuid = player.uniqueId
        uuid.restartCooldown()
        uuid.canClimb = true
        disabledPlayers.remove(uuid)
        val bossBar = BossBar.bossBar(
            "<b>Stamina".miniMsg(),
            1f,
            conf.baseBarColor,
            conf.baseOverlay
        )
        bossBar.addListener(object : BossBar.Listener {
            override fun bossBarProgressChanged(bar: BossBar, oldProgress: Float, newProgress: Float) {
                //Hide when full
                if (newProgress < 1f)
                    player.showBossBar(bar)
                else player.hideBossBar(bar)

            }
        })
        registeredBars[uuid] = bossBar
        return bossBar
    }

    @EventHandler
    fun PlayerJoinEvent.onPlayerJoin() {
        registerBar(player)
        if (player.isClimbing && player.uniqueId in barProgressTracker) {
            player.setStamina(barProgressTracker[player.uniqueId]!!)
            barProgressTracker.remove(player.uniqueId)
        } else if (player.isClimbing && player.uniqueId !in barProgressTracker) {
            ClimbBehaviour.stopClimbing(player)
        }
    }

    @EventHandler
    fun PlayerQuitEvent.onPlayerQuit() {
        val player = player
        val uuid = player.uniqueId
        if (uuid in registeredBars)
            barProgressTracker[uuid] = registeredBars[uuid]?.progress() ?: 0f
        unregisterBar(uuid)
        if (ClimbBehaviour.isClimbing.containsKey(uuid)) ClimbBehaviour.stopClimbing(player)
    }

    @EventHandler
    fun PlayerMoveEvent.onPlayerMove() {
        val uuid = player.uniqueId
        val climbDisabled = player in Tags.disabledPlayers
        val vel = player.velocity.y
        if (vel < -0.1) velocities[uuid] = vel

        if (player.isClimbing && !player.climbEnabled && !climbDisabled) Tags.disableClimb(player)

        if (player.isClimbing && !uuid.canClimb && !climbDisabled) {
            Tags.disableClimb(player)
            stamina.plugin.launch {
                while (!uuid.canClimb) delay(1.seconds)
                Tags.enableClimb(player)
            }
        }

        if (player.isClimbing && uuid.canClimb && from.distanceSquared(to) > 0.007)
            player.removeStamina(conf.staminaRemoveWhileOnClimbable)

        if (!player.isClimbing && uuid.isClimbing && from.distanceSquared(to) > 0.007)
            player.removeStamina(conf.staminaRemoveWhileMoving)
    }

    @EventHandler
    fun EntityDamageEvent.onPlayerFall() { //Remove stamina from player falls
        val player = entity as? Player ?: return
        val uuid = player.uniqueId
        if (cause != EntityDamageEvent.DamageCause.FALL || !player.climbEnabled || !velocities.containsKey(uuid)) return

        val bossBar = registeredBars[uuid] ?: return
        val threshold = 0.6
        val vel = velocities[uuid] ?: return
        player.hideBossBar(bossBar)
        if (vel > -threshold) bossBar.removeProgress(0.1f / 15f)
        else bossBar.removeProgress(damage.toFloat() / 15f) //taking 15 damage reduces stamina fully
    }

    @EventHandler
    fun PlayerDeathEvent.onPlayerDeath() {
        val player = entity
        val uuid = player.uniqueId
        if (!player.climbEnabled) return
        registeredBars[uuid]?.progress(1.0f)
    }

    @EventHandler
    fun PlayerGameModeChangeEvent.onGamemodeChange() {
        if (player.climbEnabled
            && (newGameMode == SURVIVAL || newGameMode == ADVENTURE)
            && !registeredBars.containsKey(player.uniqueId)
        ) {
            registerBar(player)
            player.setStamina(1f)
        }
    }

    fun unregisterBar(uuid: UUID) {
        ClimbBehaviour.cooldown.remove(uuid)
        val bar = registeredBars[uuid] ?: return
        Bukkit.getServer().hideBossBar(bar)
        registeredBars.remove(uuid)
    }

    inline fun forEachBar(run: (Player, UUID, BossBar) -> Unit) {
        registeredBars.toMap().forEach { (uuid, bar) ->
            val player = uuid.toPlayer() ?: registeredBars.remove(uuid).let { return@forEach }
            run(player, uuid, bar)
        }
    }

}

/** Adds [amount] progress from [bossBar]'s progress */
internal fun BossBar.addProgress(amount: Float) {
    progress((progress() + amount).coerceIn(0.0f..1.0f))
}

/** Removes [amount] progress from [bossBar]'s progress */
internal fun BossBar.removeProgress(amount: Float) {
    progress((progress() - amount).coerceIn(0.0f..1.0f))
}

/** Sets progress to [amount] from [bossBar]'s progress */
internal fun BossBar.setProgress(amount: Float) {
    progress((amount).coerceIn(0.0f..1.0f))
}

/** Adds [amount] progress from [uuid]'s associated BossBar's progress */
internal fun Player.addStamina(amount: Float) {
    StaminaBar.registeredBars[uniqueId]?.addProgress(amount)
}

/** Removes [amount] progress from [uuid]'s associated BossBar's progress */
internal fun Player.removeStamina(amount: Float) {
    StaminaBar.registeredBars[uniqueId]?.removeProgress(amount)
}

/** Sets progress to [amount] from [uuid]'s associated BossBar's progress */
internal fun Player.setStamina(amount: Float) {
    StaminaBar.registeredBars[uniqueId]?.setProgress(amount)
}
